pkg/cloudprovider/aws/instancetypes.go

```diff
@@ -0,0 +1,226 @@
+		zoneToInstanceTypeInfo[zone] = p.filterFrom(instanceTypes, constraints)
+	}
+	return zoneToInstanceTypeInfo, nil
+}
+
+// UniqueInstanceTypesFrom returns a unique slice of InstanceTypeInfo structs from the values of the instancePools passed in
+func (p *InstanceTypeProvider) UniqueInstanceTypesFrom(instancePools map[string][]*ec2.InstanceTypeInfo) []*ec2.InstanceTypeInfo {
+	result := []*ec2.InstanceTypeInfo{}
+	uniqueInstanceTypes := map[string]bool{}
+	for _, instanceTypes := range instancePools {
+		for _, instanceType := range instanceTypes {
+			if _, ok := uniqueInstanceTypes[*instanceType.InstanceType]; !ok {
+				uniqueInstanceTypes[*instanceType.InstanceType] = true
+				result = append(result, instanceType)
+			}
+		}
+	}
+	return result
+}
+
+// InstanceTypesPerZoneFrom returns a mapping of zone to InstanceTypeInfo based on the provided slice of instance types names and mapping of all zones to InstanceTypeInfo
+func (p *InstanceTypeProvider) InstanceTypesPerZoneFrom(instanceTypeNames []string, instancePools map[string][]*ec2.InstanceTypeInfo) map[string][]*ec2.InstanceTypeInfo {
+	result := map[string][]*ec2.InstanceTypeInfo{}
+	for zone, instanceTypes := range instancePools {
+		for _, instanceTypeName := range instanceTypeNames {
+			for _, instanceType := range instanceTypes {
+				if instanceType != nil && *instanceType.InstanceType == instanceTypeName {
+					result[zone] = append(result[zone], instanceType)
+				}
+			}
+		}
+	}
+	return result
+}
+
+// getAllInstanceTypes retrieves all instance types from the ec2 DescribeInstanceTypes API using some opinionated filters
+func (p *InstanceTypeProvider) getAllInstanceTypes(ctx context.Context) ([]*ec2.InstanceTypeInfo, error) {
+	instanceTypes := []*ec2.InstanceTypeInfo{}
+	describeInstanceTypesInput := &ec2.DescribeInstanceTypesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("supported-virtualization-type"),
+				Values: []*string{aws.String("hvm")},
+			},
+			{
+				Name:   aws.String("bare-metal"),
+				Values: []*string{aws.String("false")},
+			},
+			{
+				Name:   aws.String("instance-type"),
+				Values: aws.StringSlice([]string{"m5.*", "c5.*", "r5.*", "m5a.*", "c5a.*", "r5a.*", "m4.*", "c4.*", "r4.*", "m6g.*", "c6g.*", "r6g.*"}),
+			},
+		},
+	}
+	err := p.ec2.DescribeInstanceTypesPagesWithContext(ctx, describeInstanceTypesInput, func(page *ec2.DescribeInstanceTypesOutput, lastPage bool) bool {
+		instanceTypes = append(instanceTypes, page.InstanceTypes...)
+		return true
+	})
+	if err != nil {
+		return nil, fmt.Errorf("fetching instance types using ec2.DescribeInstanceTypes, %w", err)
+	}
+	return instanceTypes, nil
+}
+
+// filterFrom returns a filtered list of instance types based on the provided resource constraints
+func (p *InstanceTypeProvider) filterFrom(instanceTypes []*ec2.InstanceTypeInfo, constraints *cloudprovider.Constraints) []*ec2.InstanceTypeInfo {
+	filteredInstancePools := []*ec2.InstanceTypeInfo{}
+	architecture := "x86_64"
```

Comment:
Does this break arm support? I know Jacob had a normalize architecture function somewhere.

Reply:
no, this sets the default to x86_64 and then changes to arm64 in the next line if the constraints includes arm64. The EC2 apis support the `arm64` string so it can be set directly. 

Ah, maybe we can pull out any common aws utils into a util pkg within aws, wdut? It would feel dirty to reach into the LaunchTemplateProvider to use that normalize func.

Reply:
Sure